File: src/main/java/eu/mshade/enderchest/marshal/world/WorldBinaryTagMarshal.java
package eu.mshade.enderchest.marshal.world;

import eu.mshade.enderchest.marshal.metadata.MetadataKeyValueBinaryTagMarshal;
import eu.mshade.enderchest.world.DefaultWorld;
import eu.mshade.enderchest.world.WorldManager;
import eu.mshade.enderframe.metadata.MetadataKeyValueBucket;
import eu.mshade.enderframe.metadata.world.WorldMetadataType;
import eu.mshade.enderframe.world.World;
import eu.mshade.mwork.binarytag.BinaryTag;
import eu.mshade.mwork.binarytag.BinaryTagDriver;
import eu.mshade.mwork.binarytag.BinaryTagDynamicMarshal;
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

public class WorldBinaryTagMarshal implements BinaryTagDynamicMarshal {

    public CompoundBinaryTag serialize(BinaryTagDriver binaryTagDriver, World world){
        MetadataKeyValueBinaryTagMarshal metadataKeyValueBinaryTagMarshal = binaryTagDriver.getDynamicMarshal(MetadataKeyValueBinaryTagMarshal.class);
        return metadataKeyValueBinaryTagMarshal.serialize(WorldMetadataType.class, world.getMetadataKeyValueBucket());
    }

    public World deserialize(BinaryTagDriver binaryTagDriver, BinaryTag<?> binaryTag, WorldManager worldManager, File worldFolder){
        MetadataKeyValueBinaryTagMarshal metadataKeyValueBinaryTagMarshal = binaryTagDriver.getDynamicMarshal(MetadataKeyValueBinaryTagMarshal.class);
        MetadataKeyValueBucket metadataKeyValueBucket = metadataKeyValueBinaryTagMarshal.deserialize(WorldMetadataType.class, (CompoundBinaryTag) binaryTag);
        return new DefaultWorld(worldManager, worldFolder, metadataKeyValueBucket);
    }

    public void write(BinaryTagDriver binaryTagDriver, World world){
        try {
            FileOutputStream fileOutputStream = new FileOutputStream(new File(world.getWorldFolder(), "level.dat"));
            binaryTagDriver.writeCompoundBinaryTag(serialize(binaryTagDriver, world), fileOutputStream);
            fileOutputStream.close();
        }catch (IOException e){
            e.printStackTrace();
        }
    }

    public World read(BinaryTagDriver binaryTagDriver, File file, WorldManager worldManager){
        CompoundBinaryTag compoundBinaryTag = binaryTagDriver.readCompoundBinaryTag(new File(file, "level.dat"));
        return deserialize(binaryTagDriver, compoundBinaryTag, worldManager, file);
    }

}

File: src/main/java/eu/mshade/enderchest/marshal/world/WorldBinaryTagMarshal.kt
package eu.mshade.enderchest.marshal.world

import eu.mshade.enderframe.metadata.MetadataKeyValueBufferRegistry
import eu.mshade.enderchest.world.ChunkSafeguard
import eu.mshade.enderchest.world.DefaultWorld
import eu.mshade.enderframe.world.World
import eu.mshade.mwork.binarytag.BinaryTag
import eu.mshade.mwork.binarytag.BinaryTagDriver
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

object WorldBinaryTagMarshal {

    fun serialize(
        world: World,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ): CompoundBinaryTag {
        return metadataKeyValueBufferRegistry.serialize(world.metadataKeyValueBucket)
    }

    fun deserialize(
        binaryTag: BinaryTag<*>,
        chunkSafeguard: ChunkSafeguard,
        worldFolder: File,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ): World {
        val metadataKeyValueBucket =
            metadataKeyValueBufferRegistry.deserialize((binaryTag as CompoundBinaryTag?)!!)
        metadataKeyValueBucket.toggleTrackUpdates(true)
        return DefaultWorld(chunkSafeguard, worldFolder, metadataKeyValueBucket)
    }

    fun write(
        binaryTagDriver: BinaryTagDriver,
        world: World,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ) {
        if (world.metadataKeyValueBucket.consumeUpdatedMetadataKeyValue().isEmpty()) return
        try {
            val fileOutputStream = FileOutputStream(File(world.worldFolder, "level.dat"))
            binaryTagDriver.writeCompoundBinaryTag(
                serialize(world, metadataKeyValueBufferRegistry),
                fileOutputStream
            )
            fileOutputStream.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun read(
        binaryTagDriver: BinaryTagDriver,
        file: File,
        chunkSafeguard: ChunkSafeguard,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ): World {
        val compoundBinaryTag = binaryTagDriver.readCompoundBinaryTag(File(file, "level.dat"))
        return deserialize(compoundBinaryTag, chunkSafeguard, file, metadataKeyValueBufferRegistry)
    }
}